src/PIL/Image.py

```diff
@@ -208,43 +208,14 @@ def isImageType(t):
 else:
     _ENDIAN = ">"
 
-_MODE_CONV = {
-    # official modes
-    "1": ("|b1", None),  # Bits need to be extended to bytes
-    "L": ("|u1", None),
-    "LA": ("|u1", 2),
-    "I": (_ENDIAN + "i4", None),
-    "F": (_ENDIAN + "f4", None),
-    "P": ("|u1", None),
-    "RGB": ("|u1", 3),
-    "RGBX": ("|u1", 4),
-    "RGBA": ("|u1", 4),
-    "CMYK": ("|u1", 4),
-    "YCbCr": ("|u1", 3),
-    "LAB": ("|u1", 3),  # UNDONE - unsigned |u1i1i1
-    "HSV": ("|u1", 3),
-    # I;16 == I;16L, and I;32 == I;32L
-    "I;16": ("<u2", None),
-    "I;16B": (">u2", None),
-    "I;16L": ("<u2", None),
-    "I;16S": ("<i2", None),
-    "I;16BS": (">i2", None),
-    "I;16LS": ("<i2", None),
-    "I;32": ("<u4", None),
-    "I;32B": (">u4", None),
-    "I;32L": ("<u4", None),
-    "I;32S": ("<i4", None),
-    "I;32BS": (">i4", None),
-    "I;32LS": ("<i4", None),
-}
-
 
 def _conv_type_shape(im):
-    typ, extra = _MODE_CONV[im.mode]
-    if extra is None:
-        return (im.size[1], im.size[0]), typ
-    else:
-        return (im.size[1], im.size[0], extra), typ
+    m = ImageMode.getmode(im.mode)
+    shape = (im.size[1], im.size[0])
+    extra = len(m.bands)
+    if extra != 1:
+        shape += (extra,)
```

Comment:
```suggestion
    shape = (im.size[1], im.size[0])  # im.size[::-1]
    shape = shape if len(m.bands) == 1 else (*shape, len(m.bands))
```

Not sure if you find this syntax neater. Both options look viable to me.

Reply:
It does seem to be a matter of personal preference, but `shape = shape` looks odd to me.

I have clarified the first line slightly by changing it to `(im.height, im.width)`.

Reply:
that's fair.